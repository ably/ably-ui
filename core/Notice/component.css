@layer components {
  .ui-announcement {
    @apply font-sans;
    max-height: 37.5rem;
    transition: max-height 300ms;
  }
}


/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIndlYnBhY2s6Ly9BYmx5VWkuQ29yZS5bbmFtZV0vLi9zcmMvY29yZS9Ob3RpY2UvY29tcG9uZW50LmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQTtFQUNFO0lBQ0UsZ0JBQWdCO0lBQ2hCLG1CQUFtQjtJQUNuQiw0QkFBNEI7RUFDOUI7QUFDRiIsImZpbGUiOiJjb3JlL05vdGljZS9jb21wb25lbnQuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiQGxheWVyIGNvbXBvbmVudHMge1xuICAudWktYW5ub3VuY2VtZW50IHtcbiAgICBAYXBwbHkgZm9udC1zYW5zO1xuICAgIG1heC1oZWlnaHQ6IDM3LjVyZW07XG4gICAgdHJhbnNpdGlvbjogbWF4LWhlaWdodCAzMDBtcztcbiAgfVxufVxuIl0sInNvdXJjZVJvb3QiOiIifQ==*/