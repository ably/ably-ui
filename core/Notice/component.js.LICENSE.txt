/*!
 * JavaScript Cookie v2.2.1
 * https://github.com/js-cookie/js-cookie
 *
 * Copyright 2006, 2015 Klaus Hartl & Fagner Brack
 * Released under the MIT license
 */

/*! @license DOMPurify | (c) Cure53 and other contributors | Released under the Apache license 2.0 and Mozilla Public License 2.0 | github.com/cure53/DOMPurify/blob/2.2.2/LICENSE */
