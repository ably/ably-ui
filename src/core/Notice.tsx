import React, { ReactNode, useEffect } from "react";

import NoticeScripts from "./Notice/component.js";
import Icon from "./Icon";
type ContentWrapperProps = {
  buttonLink: string;
  children: ReactNode;
};

// TODO(jamiehenson):
// This type is a bit messed up currently due to the NoticeScripts import being interpreted as NoticeProps.
// Plan is to TS-ify the JS assets too, so this can be rectified then. The NoticeScripts-oriented props are
// the ones after the line break.
type NoticeProps = {
  buttonLink?: string;
  buttonLabel?: string;
  bodyText?: string;
  title?: string;
  closeBtn?: boolean;
  config?: {
    collapse: boolean;
    noticeId: string;
    cookieId: string;
  };
  bgColor?: string;
  textColor?: string;

  bannerContainer?: Element | null;
  cookieId?: string;
  noticeId?: string;
  options?: { collapse: boolean };
};

const contentWrapperClasses = "w-full pr-8 ui-text-p3 self-center";

const ContentWrapper = ({ buttonLink, children }: ContentWrapperProps) =>
  buttonLink ? (
    <a href={buttonLink} className={contentWrapperClasses}>
      {children}
    </a>
  ) : (
    <div className={contentWrapperClasses}>{children}</div>
  );

const Notice = ({
  buttonLink,
  buttonLabel,
  bodyText,
  title,
  config,
  closeBtn,
  bgColor = "bg-gradient-active-orange",
  textColor = "text-white",
}: NoticeProps) => {
  useEffect(() => {
    NoticeScripts({
      bannerContainer: document.querySelector('[data-id="ui-notice"]'),
      cookieId: config?.cookieId,
      noticeId: config?.noticeId,
      options: {
        collapse: config?.collapse || false,
      },
    });
  }, []);

  const wrapperClasses = ["ui-announcement", bgColor, textColor].join(" ");

  return (
    <div
      className={wrapperClasses}
      data-id="ui-notice"
      style={{ maxHeight: 0, overflow: "hidden" }}
    >
      <div className="ui-grid-px py-16 max-w-screen-xl mx-auto flex items-start">
        <ContentWrapper buttonLink={buttonLink ?? "#"}>
          <strong className="font-bold whitespace-nowrap pr-4">{title}</strong>
          <span className="pr-4">{bodyText}</span>
          {buttonLabel && (
            <span className="underline cursor-pointer whitespace-nowrap">
              {buttonLabel}
            </span>
          )}
        </ContentWrapper>

        {closeBtn && (
          <button
            type="button"
            className="ml-auto h-20 w-20 border-none bg-none self-baseline"
          >
            <Icon
              name="icon-gui-close"
              size="1.25rem"
              color="text-cool-black"
            />
          </button>
        )}
      </div>
    </div>
  );
};

export default Notice;
