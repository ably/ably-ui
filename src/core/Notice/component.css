.ui-announcement {
  @apply font-sans;
  max-height: 37.5rem;
  transition: max-height 300ms;
}
