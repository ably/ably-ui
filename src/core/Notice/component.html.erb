<%# Note the style attribute is used for entry animation %>
<%= tag.div class: "ui-announcement #{bg_color} #{text_color}", data: { id: "ui-notice" }, style: "max-height: 0; overflow: hidden" do %>
  <div class="ui-grid-px py-16 max-w-screen-xl mx-auto flex items-start">
    <%= content_wrapper class: "font-light w-full pr-8 text-p3 self-center" do %>
      <strong class="font-medium whitespace-nowrap pr-4"><%= @contents[:title] %></strong>
      <span class="pr-4"><%= @contents[:body_text] %></span>
      <span class="underline cursor-pointer whitespace-nowrap"><%= @contents[:button_label] %></span>
    <% end %>

    <% if close_btn? %>
      <%= button_tag type: 'button', class: 'h-20 w-20 mt-4 ml-auto border-none bg-none self-baseline' do %>
        <%= tag.svg class: ["h-20", "w-20", "ui-icon-white"] do %>
          <use xlink:href="#sprite-close"></use>
        <% end %>
      <% end %>
    <% end %>
  </div>
<% end %>
