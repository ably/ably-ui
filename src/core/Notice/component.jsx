import React, { useEffect } from "react";
import T from "prop-types";

const contentWrapperClasses = "font-light w-full pr-8 text-p3";

import NoticeScripts from "./component.js";

const ContentWrapper = ({ buttonLink, children }) =>
  buttonLink ? (
    <a href={buttonLink} className={contentWrapperClasses}>
      {children}
    </a>
  ) : (
    <div className={contentWrapperClasses}>{children}</div>
  );

ContentWrapper.propTypes = {
  buttonLink: T.string,
  children: T.node,
};

const Notice = ({ buttonLink, buttonLabel, bodyText, title, config, closeBtn }) => {
  useEffect(() => {
    NoticeScripts({
      bannerContainer: document.querySelector('[data-id="ui-notice"]'),
      cookieId: config.cookieId,
      noticeId: config.noticeId,
      options: {
        collapse: config.collapse,
      },
    });
  }, []);

  return (
    // Note the style attribute is used for entry animation
    <div className="ui-announcement" data-id="ui-notice" style={{ maxHeight: 0, overflow: "hidden" }}>
      <div className="ui-grid-px py-16 max-w-screen-xl mx-auto flex items-center">
        <ContentWrapper buttonLink={buttonLink}>
          <strong className="font-medium whitespace-nowrap pr-4">{title}</strong>
          <span className="pr-4">{bodyText}</span>
          {buttonLabel && <span className="underline cursor-pointer whitespace-nowrap">{buttonLabel}</span>}
        </ContentWrapper>

        {closeBtn && (
          <button type="button" className="mt-4 ml-auto border-none bg-none">
            <svg className="h-24 w-24 ui-icon-white">
              <use xlinkHref="#sprite-close"></use>
            </svg>
          </button>
        )}
      </div>
    </div>
  );
};

Notice.propTypes = {
  buttonLink: T.string,
  buttonLabel: T.string,
  bodyText: T.string,
  title: T.string,
  closeBtn: T.bool,
  config: T.shape({
    collapse: T.bool,
    noticeId: T.string,
    cookieId: T.string,
  }),
};

export default Notice;
