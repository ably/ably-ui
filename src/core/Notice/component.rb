module AblyUi
  module Core
    class Notice < ViewComponent::Base
      def initialize(contents)
        @contents = contents
      end

      def close_btn?
        @contents[:close_btn]
      end

      def content_wrapper(*args, &block)
        if @contents[:button_link].present?
          link_to(@contents[:button_link], *args, &block)
        else
          tag.div(*args, &block)
        end
      end
    end
  end
end
