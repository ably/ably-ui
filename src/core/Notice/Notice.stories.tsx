import React from "react";
import Notice from "../Notice";

export default {
  title: "Components/Notice",
  component: Notice,
  args: {
    buttonLink: "/",
    title: "Flashy title",
    bodyText: "This is a notice",
    textColor: "text-ably-primary",
    closeBtn: true,
  },
};

export const Default = {
  render: (args) => <Notice {...args} />,
};

export const RailsUjs = {
  args: {
    title: "With Rails UJS",
    bodyText:
      "<a href='/logout' data-method='delete'>Logout</a> but with script tags <script>alert('hello')</script>",
  },
};
