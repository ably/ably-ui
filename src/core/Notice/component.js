import "./component.css";
import Cookie from "js-cookie";
import throttle from "lodash.throttle";

import { queryId } from "../dom-query";
import { FLASH_DATA_ID } from "../Flash/component.tsx";

const COOKIE_EXPIRY = 90;
const COLLAPSE_TRIGGER_DISTANCE = 5;
const SCROLL_LISTENER_THROTTLE = 100;
const RESIZE_LISTENER_THROTTLE = 100;

const isMdViewport = () => !window.matchMedia(`(min-width: 65rem)`).matches;

const adjustFlashMargin = (open) => {
  // HACK ALERT
  // Add margin to flashes when opening the notice.
  // Flashes are react components and this notice needs to work as a view-component and react component.
  // We could do this with redux but then we potentially update state on every scroll event, which
  // even with debounce will deplate our frame budget.

  const flash = queryId(FLASH_DATA_ID);

  if (flash) {
    flash.style.marginTop = open ? `4rem` : null;
  }
};

const hideOnMobile = (bannerContainer) => {
  if (isMdViewport()) {
    bannerContainer.style.display = "none";
  } else {
    bannerContainer.style.display = null;
  }
};

const hideNotice = (bannerContainer) => {
  bannerContainer.style.maxHeight = 0;
  bannerContainer.style.overflow = "hidden";

  adjustFlashMargin(false);
};

const showNotice = (bannerContainer) => {
  bannerContainer.style.maxHeight = null;
  bannerContainer.style.overflow = null;

  adjustFlashMargin(true);
};

const setupRememberClosed = (cookieId, noticeId) => {
  const cookie = Cookie.get(cookieId) || "";

  Cookie.set(cookieId, `${cookie.replace(`${noticeId},`, "") + noticeId},`, {
    expires: COOKIE_EXPIRY,
  });
};

const hasBeenClosedBefore = (cookieId, noticeId) =>
  (Cookie.get(cookieId) || "").includes(noticeId);

const setupNoticeCollapse = (bannerContainer) => {
  const scrollTop = window.scrollY;

  if (scrollTop > COLLAPSE_TRIGGER_DISTANCE) {
    hideNotice(bannerContainer);
  }

  const listener = throttle(() => {
    const scrollTop = window.scrollY;

    if (scrollTop > COLLAPSE_TRIGGER_DISTANCE) {
      hideNotice(bannerContainer);
    } else if (bannerContainer.style.overflow) {
      showNotice(bannerContainer);
    }
  }, SCROLL_LISTENER_THROTTLE);

  document.addEventListener("scroll", listener);

  return () => document.removeEventListener("scroll", listener);
};

const setupCloseBtn = (
  bannerContainer,
  cookieId,
  noticeId,
  collapseUnmountListeners
) => {
  const closeBtn = bannerContainer.querySelector("button");

  if (!closeBtn) return () => {};

  const listener = () => {
    if (cookieId && noticeId) setupRememberClosed(cookieId, noticeId);

    hideNotice(bannerContainer);
    collapseUnmountListeners();
  };

  closeBtn.addEventListener("click", listener);

  return () => document.removeEventListener("click", listener);
};

const resizeHandler = (bannerContainer) => {
  const handler = throttle(() => {
    hideOnMobile(bannerContainer);
  }, RESIZE_LISTENER_THROTTLE);

  window.addEventListener("resize", handler);

  return () => window.removeEventListener("resize", handler);
};

const Notice = ({ bannerContainer, cookieId, noticeId, options }) => {
  if (typeof window === "undefined") return () => {};

  if (!bannerContainer) {
    console.warn(
      "A Notice component was initited but no notice container was found."
    );
    return () => {};
  }

  if (hasBeenClosedBefore(cookieId, noticeId)) return () => {};

  hideOnMobile(bannerContainer);
  showNotice(bannerContainer);

  const opts = { collapse: true, ...options };

  const collapseUnmountListeners = opts.collapse
    ? setupNoticeCollapse(bannerContainer)
    : () => {};

  const closeBtnUnmountListeners = setupCloseBtn(
    bannerContainer,
    cookieId,
    noticeId,
    collapseUnmountListeners
  );

  const resizeUnmountListener = resizeHandler(bannerContainer);

  return () => {
    [
      closeBtnUnmountListeners,
      collapseUnmountListeners,
      resizeUnmountListener,
    ].forEach((unmount) => unmount());
  };
};

export default Notice;
