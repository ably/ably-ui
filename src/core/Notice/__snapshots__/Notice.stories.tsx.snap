// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Notice Default smoke-test 1`] = `
<div class="ui-announcement bg-orange-100 dark:bg-orange-1100 text-neutral-1300 dark:text-neutral-000"
     data-id="ui-notice"
     style
>
  <div class="ui-grid-px py-4 max-w-screen-xl mx-auto flex items-start">
    <a href="/"
       class="w-full pr-2 ui-text-p4 self-center text-neutral-1300 dark:text-neutral-000"
    >
      <strong class="font-bold whitespace-nowrap pr-1">
        Flashy title
      </strong>
      <span class="pr-1">
        This is a notice
      </span>
    </a>
    <button type="button"
            class="ml-auto h-5 w-5 border-none bg-none self-baseline"
    >
      <svg class="text-neutral-1300 dark:text-neutral-000"
           style="width: 1.25rem; height: 1.25rem;"
      >
        <use xlink:href="#sprite-icon-gui-x-mark-outline">
        </use>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Components/Notice RailsUjs smoke-test 1`] = `
<div class="ui-announcement bg-orange-100 dark:bg-orange-1100 text-neutral-1300 dark:text-neutral-000"
     data-id="ui-notice"
     style
>
  <div class="ui-grid-px py-4 max-w-screen-xl mx-auto flex items-start">
    <a href="/"
       class="w-full pr-2 ui-text-p4 self-center text-neutral-1300 dark:text-neutral-000"
    >
      <strong class="font-bold whitespace-nowrap pr-1">
        With Rails UJS
      </strong>
      <span class="pr-1">
        <a data-method="delete"
           href="/logout"
        >
          Logout
        </a>
        but with script tags
      </span>
    </a>
    <button type="button"
            class="ml-auto h-5 w-5 border-none bg-none self-baseline"
    >
      <svg class="text-neutral-1300 dark:text-neutral-000"
           style="width: 1.25rem; height: 1.25rem;"
      >
        <use xlink:href="#sprite-icon-gui-x-mark-outline">
        </use>
      </svg>
    </button>
  </div>
</div>
`;
